evm/vm/forks/sharding/windback_worker.py

```diff
@@ -0,0 +1,212 @@
+    # chunks = DATA_SIZE // 32
+    # mtree = [0] * chunks + [data[chunks*32: chunks*32+32] for i in range(chunks)]
+    # for i in range(chunks-1, 0, -1):
+    #     mtree[i] = sha3(mtree[i*2] + mtree[i*2+1])
+    # assert mtree[i] == collation.data_root
+    return True
+
+
+def clean_done_task(tasks):
+    return [task for task in tasks if not task.done()]
+
+
+class WindbackWorker:
+
+    YIELDED_TIME = 0.1
+
+    logger = logging.getLogger("evm.chain.sharding.windback_worker")
+
+    is_time_up = None
+    latest_fetching_period = None
+
+    collation_validity_cache = None
+    chain_validity = None
+
+    unfinished_verifying_tasks = None
+    collation_verifying_task = None
+    chain_collations = None
+
+    def __init__(self, vmc, shard_tracker, my_address):
+        self.vmc = vmc
+        self.shard_tracker = shard_tracker
+        self.my_address = my_address
+
+        self.is_time_up = False
+        self.latest_fetching_period = 0
+
+        # map[collation] -> validity
+        self.collation_validity_cache = {}
+        # map[chain_head] -> validity
+        # this should be able to be updated by the collations' validity
+        self.chain_validity = defaultdict(lambda: True)
+
+        # current tasks
+        self.unfinished_verifying_tasks = []
+        # map[collation_hash] -> task
+        self.collation_verifying_task = {}
+        # map[chain_head] -> list of collations
+        # the collations in the chain
+        self.chain_collations = defaultdict(list)
+
+    def get_shard_id(self):
+        return self.shard_tracker.shard_id
+
+    # time related
+
+    def get_current_period(self):
+        return self.vmc.web3.eth.blockNumber // self.vmc.config['PERIOD_LENGTH']
+
+    # guess_head process related
+
+    async def process_collation(self, head_collation_hash, current_collation_hash, descendants):
+        collation = await download_collation(current_collation_hash)
+        collation_validity = await verify_collation(collation, current_collation_hash)
+        self.collation_validity_cache[current_collation_hash] = collation_validity
+        if not collation_validity:
+            # Set its descendants in current chain to be invalid chain heads
+            self.chain_validity[current_collation_hash] = False
+            for collation_hash in descendants:
+                self.chain_validity[collation_hash] = False
+
+    def set_collation_task(self, head_collation_hash, current_collation_hash, task):
+        self.unfinished_verifying_tasks.append(task)
+        self.collation_verifying_task[current_collation_hash] = task
+        self.chain_collations[head_collation_hash].append(current_collation_hash)
+
+    def get_chain_tasks(self, head_collation_hash):
+        # ignore tasks that are done
+        return [
+            self.collation_verifying_task[collation_hash]
+            for collation_hash in self.chain_collations[head_collation_hash]
+        ]
+
+    async def guess_head(self):
```

Comment:
I would separate `guess_head` from the windback. Windback could run indefinitely until it's cancelled, and `guess_head` could be called at any time (in the worst case it would return the genesis).